rfcs/2022-12-09-541-reload-api.md

```diff
@@ -0,0 +1,277 @@
+# RFC 541 - 2022-12-09 - Reload API
+
+This RFC proposes adding an HTTP API for Vector's configuration reloading
+process. This will allow an enterprise control plane to more directly control a
+running instance and provide a path for synchronous runtime feedback that
+doesn't currently exist.
+
+## Context
+
+- [RFC: Observability Pipelines Worker (internal)](https://docs.google.com/document/d/10c6KuzbUcEUJxBfzX3a2CQaLHGPW2zKlz-_U8kS2DmA)
+- [Observability Pipelines Worker Implementation (internal)](https://docs.google.com/document/d/1hXv0fO1nsMkZeBIgZAsGD6QB9NACYaioxQ1FoQz94LI)
+
+## Cross cutting concerns
+
+- [Reloading RFC](https://docs.google.com/document/d/1rmLwFrx_yAeLHa_SeEyJx0Op6Z0EHpK6_LVa39fnlok)
+- [Add reload subcommand](https://github.com/vectordotdev/vector/issues/541)
+- [Data Samples for OP Configuration Builder (internal)](https://docs.google.com/document/d/1L8PUzDr1G9C3ORQPvMab6lcj-mh18P1nDq5ecnPf4mE)
+
+## Scope
+
+### In scope
+
+- Integrating an API that can initiate a reload and return success or failure
+- Refactoring Vector's main application loop to better support such an API
+- Allowing configuration error messages to be passed back via the API
+
+### Out of scope
+
+- Changing the actual reloading semantics of Vector in any way
+- Implementing APIs for any other type of control actions
+
+## Pain
+
+Our enterprise control plane component needs to initate configuration reloads
+in Vector and report back success or failure. Currently, the only way to do
+that would be to manipulate the config files on disk, ensure Vector is watching
+for changes, and parse the log output to determine whether or not the reload
+succeeded. This would be difficult to implement and likely unreliable.
+
+## Proposal
+
+### User Experience
+
+At startup, Vector will look for an environment variable
+(`VECTOR_CONTROL_SOCKET_PATH`). If that variable is present and contains a
+valid path, it will start an HTTP control server listening on a Unix socket at
+that path.
+
+The control server will support a single route at `POST /config`, which will
+accept a JSON-encoded version of a Vector configuration. Similar to config
+received from a provider implementation, input globs will be expanded but
+environment variables will not be interpolated. The HTTP response will
+synchronously indicate the success or failure of the triggered reload,
+including any error messages scoped by component name.
```

Comment:
The below error response doesn't seem to be scoping the error messages by component name. I imagine there are some errors that are independent of components as well.

Reply:
Updated the example to show some structure. Again, I don't see the details of the error structure as a major decision point here. It will start simple and can be easily changed to suit the needs of the product.